### DISCUSSION TITLE
The mingw64 version does not work the same as the msvc64 version

### DISCUSSION THREAD

**Discussion by zhus** [qpdf/qpdf#1174] [General]
Hi. Is it optimisation issue, bug or some other thing? 
The same file is processed with different results.
Which version works correctly?

```
cair@PS-SRV c:\Programs
# qpdf-11.9.0-mingw64\bin\qpdf --verbose --qdf 001OK_Nasmir_HR_R_GHALIOR.pdf nul
qpdf: wrote file nul

cair@PS-SRV c:\Programs
# qpdf-11.9.0-msvc64\bin\qpdf --verbose --qdf 001OK_Nasmir_HR_R_GHALIOR.pdf nul
WARNING: 001OK_Nasmir_HR_R_GHALIOR.pdf (xref table, offset 33337124): accepting invalid xref table entry
WARNING: 001OK_Nasmir_HR_R_GHALIOR.pdf: file is damaged
WARNING: 001OK_Nasmir_HR_R_GHALIOR.pdf (xref table, offset 33337495): xref syntax invalid
WARNING: 001OK_Nasmir_HR_R_GHALIOR.pdf: Attempting to reconstruct cross-reference table
qpdf: wrote file nul
qpdf: operation succeeded with warnings; resulting file may have some problems
```

**Comment by zhus** [comment]
Tested some hardware. This happens on server with 2 cpu 6 cores each. 
May be multicpu issue.

**Comment by jberkenbilt** [comment]
There's no reason there should be a difference. Rather than writing to nul, is it possible for you to write to a separate file on each system and see if the files are the same? Are you 100% certain that the input files are identical? I've been surprised by stranger things, but right now, I'm at a loss to think of any explanation for how the two versions could both work but produce different results.

**Reply by zhus** [reply]
Thank you for participating. 
 <s>Output to real file is completly the same.</s>
The input files are not just identical, but it is only one file in current folder for both commands.
Now I have left mingw64 in production, but I feel some anxiety.
```
cair@PS-SRV c:\Programs
# qpdf-11.9.0-mingw64\bin\qpdf --verbose --qdf 001OK_Nasmir_HR_R_GHALIOR.pdf qwe1.pdf
qpdf: wrote file qwe1.pdf

cair@PS-SRV c:\Programs
# qpdf-11.9.0-msvc64\bin\qpdf --verbose --qdf 001OK_Nasmir_HR_R_GHALIOR.pdf qwe2.pdf
WARNING: 001OK_Nasmir_HR_R_GHALIOR.pdf (xref table, offset 33337124): accepting invalid xref table entry
WARNING: 001OK_Nasmir_HR_R_GHALIOR.pdf: file is damaged
WARNING: 001OK_Nasmir_HR_R_GHALIOR.pdf (xref table, offset 33337495): xref syntax invalid
WARNING: 001OK_Nasmir_HR_R_GHALIOR.pdf: Attempting to reconstruct cross-reference table
qpdf: wrote file qwe2.pdf
qpdf: operation succeeded with warnings; resulting file may have some problems
```

**Reply by zhus** [reply]
diff qwe1 and qwe2 is
```
trailer <<
  /Info 2 0 R
  /Root 1 0 R
  /Size 267
  /ID [<a6184d589b364dff9cf7bda9f840ece2><fd7b3bff34a147987be65fbda0f5db92>]
>>
startxref
55079466
%%EOF
```
and
```
trailer <<
  /Info 2 0 R
  /Root 1 0 R
  /Size 267
  /ID [<a6184d589b364dff9cf7bda9f840ece2><a6344aa3b431cb0ce76f4bff31781166>]
>>
startxref
55079466
%%EOF
```
at the end of file:
the second big number in /ID string

**Reply by zhus** [reply]
Tested on my workstation:
output files differs in the same string
but no warning about "file is damaged" appears
```
zhus@PS-WS0 c:\Programs
# qpdf-11.9.0-mingw64\bin\qpdf --verbose --qdf 001OK_Nasmir_HR_R_GHALIOR.pdf qwe1.pdf
qpdf: wrote file qwe1.pdf
```
```
zhus@PS-WS0 c:\Programs
# qpdf-11.9.0-msvc64\bin\qpdf --verbose --qdf 001OK_Nasmir_HR_R_GHALIOR.pdf qwe2.pdf
qpdf: wrote file qwe2.pdf
```

**Reply by jberkenbilt** [reply]
Difference in the second number of `/ID` is expected.

**Reply by jberkenbilt** [reply]
I can't think of why warnings would be discarded by the mingw64 version. qpdf has a thorough test suite that runs across all the versions, so if this were a systematic problem, I'm sure it would be caught and I would have heard about it. I know that's not a very helpful answer. Do you have any particular reason to favor the mingw64 version? The msvc64 version is the default because, in my informal testing, it performs slightly better, but the two versions should be functionally equivalent.

**Reply by zhus** [reply]
I think there is a problems with msvc. Here is a simplest file in which I can't find the actual xref error, but the msvc version shows a "file is corrupt" warning. This is most likely not a problem with the qpdf program. Due to a configuration error, we did not check exit code 3 (I fixed it today) and did not saw these warnings before but they are was a flow. I'll try to do some research tomorrow. Or is it a problem with the operating system -- it is very ancient. Thank you!
[test32_4_.pdf](https://github.com/qpdf/qpdf/files/14563970/test32_4_.pdf)
[test32_4_.pdf`fail.txt](https://github.com/qpdf/qpdf/files/14563972/test32_4_.pdf.fail.txt)

**Comment by jberkenbilt** [comment]
Also, qpdf is completely single-threaded, so multicpu shouldn't be a factor.
